zernikeprocess.cpp

```diff
@@ -848,7 +848,7 @@ void zernikeProcess::fillVoid(wavefront &wf){
                     if (x < 0 || y < 0 || x >= wf.data.cols || y >= wf.data.rows){
                         continue;
                     }
-                    if (wf.mask.at<bool>(y,x) == 0.){
+                    if (wf.mask.at<uchar>(y,x) == 0.){
```

Comment:
are you sure about this line ?
Because :
https://github.com/githubdoe/DFTFringe/blob/1daf63a5dba50d9512b41a11263c122a84df85af/wavefront.h#L31

I'm doing C++ only for 1 year and not full time so my understanding might be wrong but shouldn't it be :
```suggestion
                    if (wf.mask.at(y,x) == false){
```

or maybe 
```suggestion
                    if (wf.mask.at<CV_U8>(y,x) == 0.){
```

For me the mask and workMask are managed in a inconsistent way unless I miss something about C++ or `cv::Mat_`. They are matrix of bools ? Then why do they store different values than true and false ? 
https://github.com/githubdoe/DFTFringe/blob/1daf63a5dba50d9512b41a11263c122a84df85af/surfacegraph.cpp#L263

Reply:
This line of code was very frustrating and took me a few hours to get working (this was a month or two ago when I was doing a similar edit).  It would work fine in release but when I stepped through with the debugger it wasn't working.  I tried changing the "0." to "false" and that didn't help.  I finally just copied code that Dale had elsewhere (the <uchar> thing) and it worked so I left it.  I did not try leaving out the "type" and I did not try "CV_U8".

Why does C++ use a whole byte for booleans?  I think it's more efficient (speed - not memory size of course).  It's quite ugly to assume that a bool is a whole byte long so I try to never do that but on mac/linux/pc it will always by a byte.  On some future platform, who knows?  I don't think c++ standard specifies that it is exactly one byte.  I'm not sure.

Reply:
Regarding the "if" statements.  My actual perferred way to do this is to make it more readable than more bug free.  A strange preference!  I know!  

I have been reading other people's code since I was 10 years old I value readability the most.

My actual preferred way would be to put each command on the same line as the if statement (not always - just in this specific code example) so more like:
if (a<0) a=0;

I've seen people do:
a=max(a,0);
but find that very difficult to read as people get negative values and magnitudes confused.  I just think it's harder to read the min() and max() method.

I know you like braces around a single line if statement as you mentioned it also a day or so ago but I don't. Tabbing is more important.  A good editor or compiler should warn me if the tabs are off.  Right now there are too many warnings but fortunately you may be fixing that.

I did think about using the "else" statements.  That should in theory improve the efficiency of the code but I felt the gain in efficiency was too small to be worth hurting the readability somewhat.

Reply:
Believe me, I think about style and readability and possible bugs all the time.  One thing I do is think "if I were the boss, would I tell this programmer to change this?" and if the answer is yes then I change it.  I don't always remember to use that trick but usually.  Always there is this voice "you should add braces", "you should use a better variable name" and so on.  I usually listen to the voice and "do it the right way".

Reply:
Back to the boolean - I tried other things as well.  I tried "== false"  "!= true".  Eventually I just copied Dale's other code that was working.  Maybe bool is more than one byte and it was indexing to the wrong location of the matrix.  But then why is bool a different sizeof() value in debug versus release?

Reply:
boolean is indeed undefined https://en.cppreference.com/w/cpp/language/types

I did a bit more research after the comment and my best guess would be `wf.mask.at(y,x) == false` without the static cast and with false. However I did not debug through the code, so if you tell me what you did is working I believe you. I would just prefer to understand why it works.

Additional guess: the mat should contain bool but because of poor management elsewhere it contains other values than true or false that mess things up ?

--------

About the braces, I also prefer the one liner you propose in comment. Really this is matter of taste, I won't push for this change.   
About the `else if` however, I find it more logical this way and not only for improving performances. Also matter of taste.

For sure I know I'm very picky on naming and other things, my colleagues are too :) I just need some time to adapt to you as my peer.

Reply:
+1 About the bool, while explaining to @hish15 at work, `wf.mask.at<uchar>(y,x) == 0.` this is comparing a uchar to a double 0. ? I'm not sure what this does.  https://godbolt.org/ tells it's the same as compare to int 0 but this little dot might explain the difference between debug and release ?
still unclear what happens

Reply:
I'll test this today.  This is interesting.  The period indeed is ugly as it makes it a double.  I almost deleted the period but was just trying to get this working and was frustrated at the time (again - was a few months ago).

But now I want to try your ideas.  You may run into this or related code when you fix all the warnings.  It's possible the other code where the matrix was loaded will have a warning.

Reply:
Is the <uchar> a static cast?  I thought it was as template function?  I still think it's a template function but didn't look at the definition of the "at" member function.

Anyway I'll test the code today.

Reply:
in vectors/arrays, `at` is same as `[]` but is supposed to make boundary check. What is done in openCV I don't know.

My local expert @Hish15 :) explained me this `<>` in indeed not a cast (I wronly though it was as https://docs.opencv.org/3.4.15/df/dfc/classcv_1_1Mat__.html#a098ec36001b3998a7511e3c9169f93c8) 
But here it is a template specification. So here it should be either bool (with respect of declaration https://github.com/githubdoe/DFTFringe/blob/1daf63a5dba50d9512b41a11263c122a84df85af/wavefront.h#L31) OR nothing. But anything different from bool or nothing is not clean and might even be UB (undefined behavior).

so the only "correct" syntax would be
`wf.mask.at<bool>(y,x) == false`
or
`wf.mask.at(y,x) == false`